### PR TITLE
Always send new healthy endpoints

### PR DESCRIPTION
**Pull request by dddddai**
**Please provide a description of this PR:**
We should always send new healthy endpoints. This typically happens when EndpointSlices are manually managed or pods are selected by a ServiceEntry
Fixes https://github.com/istio/istio/issues/48373

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/model/endpointshards.go [hunk 1]
```diff
@@ -318,10 +318,11 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
 				newIstioEndpoints = append(newIstioEndpoints, nie)
 			} else {
 				// If the endpoint does not exist in shards that means it is a
-				// new endpoint. Always send new endpoints even if they are not healthy.
+				// new endpoint. Always send new healthy endpoints.
+				// Also send new unhealthy endpoints when SendUnhealthyEndpoints is enabled.
 				// This is OK since we disable panic threshold when SendUnhealthyEndpoints is enabled.
-				// Without SendUnhealthyEndpoints we do not need this; headless services will trigger the push in the Kubernetes controller.
-				if features.SendUnhealthyEndpoints.Load() {
+				// For headless services, the Kubernetes controller will trigger the push .
```
Comment (hzxuzhonghu) [inline pilot/pkg/model/endpointshards.go:324]: I cannot see how this is related to headless service

Comment (dddddai) [reply pilot/pkg/model/endpointshards.go:324]: I don't think it's related, just to preserve the comment

Comment (hzxuzhonghu) [reply pilot/pkg/model/endpointshards.go:324]: Looks  out of date

**Review by howardjohn — APPROVED** [review]
cc @ramaraochavali

**Comment by hzxuzhonghu** [comment]
need back port?

**Comment by dddddai** [comment]
> need back port?

Yes

File: pilot/pkg/model/endpointshards.go [hunk 2]
```diff
@@ -318,10 +318,10 @@ func (e *EndpointIndex) UpdateServiceEndpoints(
 				newIstioEndpoints = append(newIstioEndpoints, nie)
 			} else {
 				// If the endpoint does not exist in shards that means it is a
-				// new endpoint. Always send new endpoints even if they are not healthy.
+				// new endpoint. Always send new healthy endpoints.
+				// Also send new unhealthy endpoints when SendUnhealthyEndpoints is enabled.
 				// This is OK since we disable panic threshold when SendUnhealthyEndpoints is enabled.
-				// Without SendUnhealthyEndpoints we do not need this; headless services will trigger the push in the Kubernetes controller.
-				if features.SendUnhealthyEndpoints.Load() {
+				if nie.HealthStatus != UnHealthy || features.SendUnhealthyEndpoints.Load() {
```
[the remark this answers is not in the document]
Comment (dddddai) [reply pilot/pkg/model/endpointshards.go:324]: I agree `nie.HealthStatus == Healthy` is more clear. My idea was to make it aligned with:https://github.com/istio/istio/blob/9534b78b5bcbb61f0f44444f290d86ca1e2f369a/pilot/pkg/xds/endpoints/endpoint_builder.go#L496

I don't mind change this to `nie.HealthStatus == Healthy` though

[the remark this answers is not in the document]
Comment (freedomljc) [reply pilot/pkg/model/endpointshards.go:324]: I think the purposes of these two parts are different, and probably we dot need to make them aligned.  The code snippet in endpoint_builder is to determine whether to include unhealthy endpoint in EDS, whereas this code snippet here is to determine whether we should have EDS pushed based on endpoint changes. 

For new endpoint, there should be only two states: healthy and unhealthy. I think there is no need to consider draining status here, it should be fine and more accurate if changing it to `nie.HealthStatus == Healthy`.

**Comment by hzxuzhonghu** [comment]
1.18 is EOL

**Comment by hanxiaop** [comment]
> 1.18 is EOL

We've already cherry-picked into the release-1.20.

**Pull request merged**